src/main/java/com/couchbase/lite/auth/OpenIDConnectAuthorizer.java

```diff
@@ -0,0 +1,331 @@
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Member variables
+    ////////////////////////////////////////////////////////////
+
+    private String username;                     // SG user name
+
+    // internal
+    protected OIDCLoginCallback loginCallback;    // App-provided callback to log into the OP
+    protected ITokenStore tokenStore;
+
+    protected URL authURL;                        // The OIDC authentication URL redirected to by the OP
+    protected String IDToken;                     // Persistent ID token, when logged-in
+    protected String refreshToken;                // Persistent refresh token, when logged-in or refreshing login
+    protected boolean haveSessionCookie;          // YES if the server set a login session cookie
+
+    ////////////////////////////////////////////////////////////
+    // Constructors
+    ////////////////////////////////////////////////////////////
+    public OpenIDConnectAuthorizer(OIDCLoginCallback callback, ITokenStore tokenStore) {
+        this.loginCallback = callback;
+        this.tokenStore = tokenStore;
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Public methods
+    ////////////////////////////////////////////////////////////
+    @Override
+    public String toString() {
+        return String.format(Locale.ENGLISH, "OpenIDConnectAuthorizer[%s]", remoteURL);
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Implementations of ICustomHeadersAuthorizer
+    ////////////////////////////////////////////////////////////
+
+    // Auth phase (when we have the ID token):
+
+    @Override
+    public boolean authorizeURLRequest(Request.Builder builder) {
+        loadTokens();
+        if (IDToken != null && !haveSessionCookie) {
+            String auth = String.format(Locale.ENGLISH, "Bearer ", IDToken);
+            builder.addHeader("Authorization", auth);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Implementations of ISessionCookieAuthorizer(ILoginAuthorizer)
+    ////////////////////////////////////////////////////////////
+
+    // #pragma mark - LOGIN:
+
+    @Override
+    public List<Object> loginRequest() {
+        loadTokens();
+
+        // If we got here, 'GET _session' failed, so there's no valid session cookie or ID token.
+        IDToken = null;
+        haveSessionCookie = false;
+
+        String path;
+        if (refreshToken != null)
+            path = String.format(Locale.ENGLISH, "_oidc_refresh?refresh_token=%s",
+                    URIUtils.encode(refreshToken));
+        else if (authURL != null)
+            path = String.format(Locale.ENGLISH, "_oidc_callback?%s",
+                    authURL.getQuery());
+        else
+            path = "_oidc_challenge";
```

Comment:
I believe you would need to append `?offline=true` to the path in order to get refresh token.